phpBB/phpbb/db/migration/schema_generator.php

```diff
@@ -84,171 +98,238 @@ public function get_schema()
 		}
 
 		$migrations = $this->class_names;
+		$filter = function($class_name) {
+			return !migrator::is_migration($class_name);
+		};
+
+		$edges = function($class_name) {
+			return $class_name::depends_on();
+		};
 
-		$tree = array();
-		$check_dependencies = true;
-		while (!empty($migrations))
+		$apply_for_each = function($class_name) {
+			$this->apply_migration_to_schema($class_name);
+		};
+
+		try
+		{
+			topologicalSort($migrations, $edges, true, $apply_for_each, $filter);
+		}
+		catch (LogicException $e)
 		{
-			foreach ($migrations as $key => $migration_class)
+			throw new UnexpectedValueException(
+				"Migrations either have circular dependencies or unsatisfiable dependencies."
```

Comment:
`Migrations have circular dependencies or unfulfilling dependencies.` ?

Reply:
or `unsatisfactory`

Reply:
unsatisfiable is a maths term. It is as precise as you can get.

Reply:
Do you mean numerology? 3 or 7?

Reply:
I mean at this point the graph either has Circular Dependencies, hence not a DAG and as such cannot be sorted, or the graph is not well formed, as in it has edges to non-existent nodes (e.g. dependencies to migrations which do not exist).